includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php

```diff
@@ -87,43 +92,49 @@ protected function limit_length( $string, $limit = 127 ) {
+			'zip'           => $this->limit_length( wc_format_postcode( $order->get_billing_postcode(), $order->get_billing_country() ), 32 ),
+			'country'       => $this->limit_length( $order->get_billing_country(), 2 ),
+			'email'         => $this->limit_length( $order->get_billing_email() ),
+		);
+		$phone_number_args = $this->get_phone_number_args( $order );
+		$shipping_args     = $this->get_shipping_args( $order );
+		$query_str_except_line_items = http_build_query( array_merge( $initial_paypal_args, $phone_number_args, $shipping_args ), '', '&' );
+		$url_except_line_items_length = strlen( $this->request_url_prefix . $query_str_except_line_items );
+
 		return apply_filters(
 			'woocommerce_paypal_args', array_merge(
-				array(
-					'cmd'           => '_cart',
-					'business'      => $this->gateway->get_option( 'email' ),
-					'no_note'       => 1,
-					'currency_code' => get_woocommerce_currency(),
-					'charset'       => 'utf-8',
-					'rm'            => is_ssl() ? 2 : 1,
-					'upload'        => 1,
-					'return'        => esc_url_raw( add_query_arg( 'utm_nooverride', '1', $this->gateway->get_return_url( $order ) ) ),
-					'cancel_return' => esc_url_raw( $order->get_cancel_order_url_raw() ),
-					'page_style'    => $this->gateway->get_option( 'page_style' ),
-					'image_url'     => esc_url_raw( $this->gateway->get_option( 'image_url' ) ),
-					'paymentaction' => $this->gateway->get_option( 'paymentaction' ),
-					'bn'            => 'WooThemes_Cart',
-					'invoice'       => $this->limit_length( $this->gateway->get_option( 'invoice_prefix' ) . $order->get_order_number(), 127 ),
-					'custom'        => wp_json_encode(
-						array(
-							'order_id'  => $order->get_id(),
-							'order_key' => $order->get_order_key(),
-						)
-					),
-					'notify_url'    => $this->limit_length( $this->notify_url, 255 ),
-					'first_name'    => $this->limit_length( $order->get_billing_first_name(), 32 ),
-					'last_name'     => $this->limit_length( $order->get_billing_last_name(), 64 ),
-					'address1'      => $this->limit_length( $order->get_billing_address_1(), 100 ),
-					'address2'      => $this->limit_length( $order->get_billing_address_2(), 100 ),
-					'city'          => $this->limit_length( $order->get_billing_city(), 40 ),
-					'state'         => $this->get_paypal_state( $order->get_billing_country(), $order->get_billing_state() ),
-					'zip'           => $this->limit_length( wc_format_postcode( $order->get_billing_postcode(), $order->get_billing_country() ), 32 ),
-					'country'       => $this->limit_length( $order->get_billing_country(), 2 ),
-					'email'         => $this->limit_length( $order->get_billing_email() ),
-				),
-				$this->get_phone_number_args( $order ),
-				$this->get_shipping_args( $order ),
-				$this->get_line_item_args( $order )
+				$initial_paypal_args,
+				$phone_number_args,
+				$shipping_args,
+				$this->get_line_item_args( $order, $url_except_line_items_length )
```

Comment:
I wonder if instead of generating the args twice and passing sizes around we should instead change `get_line_item_args` to allow you to force it to a single item.

That way you can grab these args, and filter them (so extensions can make changes) THEN test length. If it's too long, call this and `get_line_item_args` again but this time force it to a single item?

Reply:
You're absolutely right, the length should be tested AFTER the filter, fixing it now.

However, I can't see if we can avoid running the filter and generating the args twice. 
Silly example: if the filter changed e.g one of shipping args based on number of line items, then we have to generate the args and run the filter again with one line item. 
Or am I missing something?

The change to make `get_line_item_args` accept the force single item parameter is much nicer, I completely agree, fixed that.